src/backends/perf_pt/collect.c

```diff
@@ -142,6 +163,84 @@ bool perf_pt_start_tracer(struct tracer_ctx *, struct perf_pt_trace *, struct pe
 bool perf_pt_stop_tracer(struct tracer_ctx *tr_ctx, struct perf_pt_cerror *);
 bool perf_pt_free_tracer(struct tracer_ctx *tr_ctx, struct perf_pt_cerror *);
 
+
+/*
+ * Reads samples from the data buffer and acts upon them appropriately.
+ *
+ * Returns true on success, or false otherwise.
+ */
+static bool
+handle_sample(void *aux_buf, struct perf_event_mmap_page *hdr,
+         struct perf_pt_trace *trace, void *data_tmp,
+         struct perf_pt_cerror *err)
+{
+    bool ret = true;
+
+    // We use atomic loads and stores with orderings to avoid concurrency
+    // issues WRT data buffer state mutably shared between us and the kernel.
+    // Without these, for example the compiler might be able to load data
+    // from `data_buf` before reading from the `hdr->data_head`.
+    void *data = (void *) hdr + hdr->data_offset;
+    __u64 head_monotonic =
+            atomic_load_explicit((_Atomic __u64 *) &hdr->data_head,
+                                 memory_order_acquire);
```

Comment:
Given that this value doesn't appear to have a dependency on anything else, shouldn't this be a relaxed load? Or is there some sort of implicit ordering with something else?

Reply:
As I understand it, we don't want the load of the buffer to come before the load of the head. It's the same logic as in `read_aux`.

Reply:
I don't remember `read_aux` :) Still, this is another chance to think about this again.

Right, so there's a dependency between `hdr->data_head` and `data` right? I presume that only PT changes `hdr->data_head` but only we change `hdr->data_tail`?

Reply:
correct.

Reply:
OK, can we note this down in a comment? Because now I understand why you have acquire/release loads.

Reply:
I'm not sure it's necessary as it's well documented in the `perf_event_open` manual. If you feel strongly, OK.

Reply:
Given that it's over 2000 lines long, that man page may also contain the secret to eternal life...

As you know, in general, I'm not a fan of too many comments. But anything to do with atomics has to be treated with the utmost caution. The problem is that the programmer's intent is not generally obvious from the code (I had to guess the above). And then I'd need to know what to look for in the man page to see the comment `On  SMP-capable  platforms,  after  reading the data_head value, space should issue an rmb().`.

Reply:
Ah, well that part is already mentioned in a comment. I thought you meant you wanted to add:
> only PT changes hdr->data_head but only we change hdr->data_tail

Reply:
Currently we have:
```
      // We use atomic loads and stores with orderings to avoid concurrency          
      // issues WRT data buffer state mutably shared between us and the kernel.   
      // Without these, for example the compiler might be able to load data        
      // from `data_buf` before reading from the `hdr->data_head`.             
```

Does that cover it for you? If not, how should we change it to be clearer?

Reply:
It's not really sufficient. For example, it doesn't say what the dependencies we have to observe actually are: it has just a very vague line about what the compiler could do (who cares what it could do? I only care about the *bad* things it could do!).

I would prefer to be clear here with something like: "There is a dependency between hdr->data_head (which is only written to by PT) and data (whose contents are also written to by PT): we need to use an acquire barrier to ensure that the former is a valid pointer into the latter. Although hdr->data_tail is only written to by us, since we may have more than one of our own threads writing to it, we also need to use an atomic to ensure that is synchronised between threads."

BTW, writing that out makes me realise that we probably only need "relax" ordering on `hdr->tail`...

Reply:
> since we may have more than one of our own threads writing to it

That would cause chaos. I don't think we can trace the same tid twice.

> we need to use an acquire barrier to ensure that the former is a valid pointer

The pointer will be valid, but maybe not up to date.

I'm not sure we are agreeing. As I see it there's a dependency between data and *both* head and tail. We want to make sure that we always have up to date values of those before looking into the data buffer at those offsets.

Reply:
> BTW, writing that out makes me realise that we probably only need "relax" ordering on hdr->tail

I don't understand, can you explain?

Reply:
Because we are the only thread to write the tail, we can rely on program order to get it right?

Reply:
>That would cause chaos. I don't think we can trace the same tid twice.

I'm not saying that we trace the same TID twice: I'm saying that, I think, thread t1 and thread t2 might read from the PT buffer at different times.

>As I see it there's a dependency between data and both head and tail.

AFAICS *we* never write to `data` or `hdr->data_head`, and we are fully in control of `hdr->data_tail`. If so, there's no write dependency between the first two and the third.

Reply:
> I'm not saying that we trace the same TID twice: I'm saying that, I think, thread t1 and thread t2 might read from the PT buffer at different times.

Don't worry about this. If we are tracing two different TIDs then t1 and t2 would each have their own mmaped buffers. Otherwise the packets would interleave and you wouldn't know which packet is for which thread.

> AFAICS we never write to data or hdr->data_head, and we are fully in control of hdr->data_tail. If so, there's no write dependency between the first two and the third.

I'm confused.

Reply:
Maybe I understand this.

Reply:
Wait, the kernel still has to read the tail so as to know not to overwrite data past this point. The release causes the other cores to see the new value at the time they acquire the tail.

Reply:
Then you still only need relaxed ordering, since `tail` stands on its own (we haven't written to `head` or `data`).

Reply:
I'm totally confused.

I suggest we keep it as is, as it's more conservative. You can try and explain it to me at the next thrash ;)

Reply:
My suggestion is the following:

```
let head = atomic_read(&hdr->data_head, ACQUIRE);
let tail = atomic_read(&hdr->data_tail, RELAXED);
...
<do the memcpy stuff>
...
atomic_write(&hdr->data_tail, head, RELAXED);
```

We have to use ACQUIRE for `hdr->data_head` because we want to make sure that everything in `data` is upto date (and the kernel's write to `hdr->data_head` serves a sort-of lock to guarantee that). However, we can get away with RELAXED for `hdr->data_tail` because we're the only people to write to it (and it doesn't need a dependency on `data`).

Feel free to challenge me on this!

Reply:
What I don't get is that the atomic read will get the latest version of the head pointer regardless of the ordering. AFAICS the ordering just changes where the read could happen. We know that the atomic read must happen before `if (tail <= head) {` because it uses `head`. So why would it matter if we load the tail or the buffer size before the head? They are the only remaining memory operations left which could be reordered. Do you follow?

Reply:
(All uses of `data` come after `if (tail <= head)`)

Reply:
The point is we want everything we read from `data` to be up to date. If we read `head` before everything to `data` is flushed, we might be in trouble.

Reply:
Sorry if I'm being dumb, but why does reading head with acquire order guarantee the data buffer to be flushed?

Reply:
If we assume that the kernel uses the right memory ordering to ensure that the head pointer is only updated after the data is written, then any atomic read of the head which occurs prior to the memcpy should give the correct result.

Reply:
Both sides have to use the right ordering (and, I admit, I'm being pessimistic: I don't know *when* the kernel updates `head`, so I'm assuming the worst case). If they used release but we used relaxed, we would read `tail` correctly but maybe not `data`.

Reply:
I think it's safe to assume the kernel does it right.

Reply:
Looking at the perf source code, this is how they read the head and write the tail:
https://github.com/torvalds/linux/blob/3be4aaf4e2d3eb95cce7835e8df797ae65ae5ac1/tools/perf/util/mmap.h#L71-L88

So here they are using not atomics, but barriers.

The meanings of those barriers are:

>mb() | A full  system memory barrier. All memory operations before the mb() in the  instruction stream will be committed before any operations after the  mb() are committed. This ordering will be visible to all bus masters in  the system. It will also ensure the order in which accesses from a  single processor reaches slave devices.

> rmb() | Like  mb(), but only guarantees ordering between read accesses. That is, all  read operations before an rmb() will be committed before any read  operations after the rmb().

I don't know that these barriers map well to the ordered atomic ops that we want to use though.

Reply:
It seems we are the only one on the entire of github using C11 atomics on the perf buffers :\

Reply:
I think I can offer reasoning as to why the code in-tree is safe.

There are two things we need to prevent:
 * Reading the data buffer before the most recent head is obtained. This would mean that we may read nothing when there is really data available.
 * Updating the tail before we have read data out of the data buffer. This would mean the kernel is free to overwrite the data we have not read yet.

Then:
 * By reading the head first in acquire order, it means we simply cannot read the data prior to holding the latest head value.
 * By writing the tail with release ordering, it means the memcpys cannot be moved after the write to the tail.

Reply:
The thing is, we don't know when head and data are written with respect to each other. What happens if head is written first and then data is updated? [I think it's unlikely that this is the way things are done, but no document seems to say it couldn't happen, so we have to be conservative.]

Reply:
The above says nothing about what ordering should be used for the one remaining atomic, the load of the tail.

I think since we are the only writer of the tail, the ordering doesn't matter. The read just has to happen before the write, which is a given for the same atomic variable, and thus can be relaxed.

(elided)

```
    void *data = (void *) hdr + hdr->data_offset;
    __u64 head_monotonic =
            atomic_load_explicit((_Atomic __u64 *) &hdr->data_head,
                                 memory_order_acquire);
    __u64 size = hdr->data_size; // No atomic load. Not mutated in-kernel.
    __u64 head = head_monotonic % size; // Head must be manually wrapped.
    __u64 tail = atomic_load_explicit((_Atomic __u64 *) &hdr->data_tail,
                                 memory_order_relaxed); // <<<<< CHANGED ORDER

    // Copy samples out, removing wrap in the process.
    void *data_tmp_end = data_tmp;
    if (tail <= head) {
        // Not wrapped.
        memcpy(data_tmp, data + tail, head - tail);
        data_tmp_end += head - tail;
    } else {
        // Wrapped.
        memcpy(data_tmp, data + tail, size - tail);
        data_tmp_end += size - tail;
        memcpy(data_tmp + size - tail, data, head);
        data_tmp_end += head;
    }
atomic_store_explicit((_Atomic __u64 *) &hdr->data_tail, head, memory_order_release);
```

Do you agree?

Reply:
> What happens if head is written first and then data is updated?

That's called a kernel bug. And if you really are worried I can find the code responsible in the kernel.

Reply:
Both sides have to cooperate or neither side can win.

Reply:
>That's called a kernel bug. And if you really are worried I can find the code responsible in the kernel.

Why is this a bug?

Reply:
Of course it could work that way. It would be odd, but unless the spec disallows it, one can't assume it will never happen. Indeed, unless the spec says there's an ordering between head and data, we can't assume anything. Hence why a full barrier is the only safe option. [I'm not saying I like this outcome, but I think it's the one that's least likely to bite us in the rear end in the future.]

Reply:
> Of course it could work that way.

I don't follow. How would you know if it was safe to read the data behind a mutex if the value of a mutex had no bearing on if the data had been written?

Reply:
There is no mutex here though.

Reply:
Its the same. How would you know how much data to read from the buffer if the kernel didn't guarantee that all data up to `head` was ready? That would be a logic bug and there's nothing the user thread could ever do to be correct under that scenario.

Reply:
I've pushed a change with a big comment explaining my reasoning. Does it work for you?

Reply:
Sort-of, but only because the man page says:
```
       data_head
              This points to the head of the data section.  The value continu‐
              ously  increases, it does not wrap.  The value needs to be manu‐
              ally wrapped by the size of the mmap buffer before accessing the
              samples.

              On  SMP-capable  platforms,  after  reading the data_head value,
              user space should issue an rmb().
```
And because it doesn't say anything like that for data_tail. In other words, the *only* reason this is correct is because the spec says it is. I think I would prefer to point the reader at the spec rather than engage in reasoning about the system's internals.

[FWIW, I think the docs for this are terrible. They could have saved us much heartache if they'd documented stuff correctly. Oh well.]

Reply:
What spec are we talking about here? The C11 language spec?

I'm actually very happy with the comment I've written as I feel it's the first time I've understood this stuff properly. I feel that in a 6 months time I'd be able to understand based on this comment.